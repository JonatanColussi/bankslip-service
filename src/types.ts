import { BANKSLIP_TYPES } from './constants';

export type BankslipTypes = keyof typeof BANKSLIP_TYPES;

export interface Bankslip {
  amount?: number;
  barCode?: string;
  expirationDate?: string;
  setAmount(): void;
  setBarcode(): void;
  setExpirationDate(): void;
  type?: BankslipTypes;
  validate(): void;
}
