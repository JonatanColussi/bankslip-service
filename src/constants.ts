export const BANKSLIP_TYPES = {
  TITULO: 'TITULO' as const,
  CONVENIO: 'CONVENIO' as const,
};
