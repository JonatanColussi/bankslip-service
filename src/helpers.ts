import { APIGatewayProxyResult } from 'aws-lambda';

export function response(data: Record<string, any>, statusCode?: number) {
  return Promise.resolve({
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'Application/json',
    },
    statusCode: statusCode || 200,
  } as APIGatewayProxyResult);
}
