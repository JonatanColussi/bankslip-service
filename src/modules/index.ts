export * from './bankslip';
export * from './convenio';
export * from './titulo';
