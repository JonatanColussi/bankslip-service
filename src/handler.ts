import { APIGatewayProxyEvent, APIGatewayProxyResult, Handler } from 'aws-lambda';

import { BANKSLIP_TYPES } from './constants';
import { response } from './helpers';
import { BankslipBase, Convenio, Titulo } from './modules';

export const viewBankslipInfo: Handler<APIGatewayProxyEvent, APIGatewayProxyResult> = event => {
  try {
    const { digitableLine = '' } = event.pathParameters || {};
    const { type } = new BankslipBase(digitableLine);

    const {
      amount = null,
      barCode,
      expirationDate = null,
    } = type === BANKSLIP_TYPES.CONVENIO ? new Convenio(digitableLine) : new Titulo(digitableLine);

    return response({
      amount,
      barCode,
      expirationDate,
    });
  } catch (error: any) {
    return response({ message: error?.message }, 400);
  }
};
