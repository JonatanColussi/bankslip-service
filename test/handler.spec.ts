import createEvent from '@serverless/event-mocks';

import { viewBankslipInfo } from '../src/handler';

it.each([
  '21290001192110001210904475617405975870000002000',
  '826100000015165011000008010000870633442022069557',
  '33690000090000001004100731323143190290002085742',
  '03399492813698200090709545001027500000000000000',
  '00190000090250038600400684863178390220000048550',
  '836100000055483500863343933736760093100258416914',
  '846500000001749900791007011198559871921070373342',
  '848900000002404201622015806051904292586034111220',
  '858000000070438403281922630720192528304729600523',
])('should parse digitableLine properly', async digitableLine => {
  const response = await viewBankslipInfo(
    createEvent('aws:apiGateway', {
      pathParameters: { digitableLine },
    } as any),
    {} as any,
    {} as any,
  );

  if (!response) {
    return;
  }

  expect(response.statusCode).toBe(200);
  expect(JSON.parse(response.body)).toMatchSnapshot();
});

it.each(['1234567890', 'abcde', 'AB1234', undefined])(
  'should return error if digitable line is invalid',
  async digitableLine => {
    const response = await viewBankslipInfo(
      createEvent('aws:apiGateway', {
        pathParameters: { digitableLine },
      } as any),
      {} as any,
      {} as any,
    );

    if (!response) {
      return;
    }

    expect(response.statusCode).toBe(400);
    expect(JSON.parse(response.body)).toEqual({
      message: 'The digitable line must contains only numbers',
    });
  },
);

it('should return an error if call handler with empty event', async () => {
  const response = await viewBankslipInfo({} as any, {} as any, {} as any);

  if (!response) {
    return;
  }

  expect(response.statusCode).toBe(400);
  expect(JSON.parse(response.body)).toEqual({
    message: 'The digitable line must contains only numbers',
  });
});
