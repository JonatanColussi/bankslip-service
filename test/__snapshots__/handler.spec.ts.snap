// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse digitableLine properly 1`] = `
Object {
  "amount": 20,
  "barCode": "21299758700000020000001121100012100447561740",
  "expirationDate": "2018-07-16",
}
`;

exports[`should parse digitableLine properly 2`] = `
Object {
  "amount": 116.5,
  "barCode": "82610000001165011000000100008706344202206955",
  "expirationDate": null,
}
`;

exports[`should parse digitableLine properly 3`] = `
Object {
  "amount": 20857.42,
  "barCode": "33691902900020857420000000000010040073132314",
  "expirationDate": "2022-06-27",
}
`;

exports[`should parse digitableLine properly 4`] = `
Object {
  "amount": null,
  "barCode": "03395000000000000009492836982000900954500102",
  "expirationDate": null,
}
`;

exports[`should parse digitableLine properly 5`] = `
Object {
  "amount": 485.5,
  "barCode": "00193902200000485500000002500386000068486317",
  "expirationDate": "2022-06-20",
}
`;

exports[`should parse digitableLine properly 6`] = `
Object {
  "amount": 548.35,
  "barCode": "83610000005483500863349337367600910025841691",
  "expirationDate": null,
}
`;

exports[`should parse digitableLine properly 7`] = `
Object {
  "amount": 74.99,
  "barCode": "84650000000749900791000111985598792107037334",
  "expirationDate": null,
}
`;

exports[`should parse digitableLine properly 8`] = `
Object {
  "amount": 40.42,
  "barCode": "84890000000404201622018060519042958603411122",
  "expirationDate": null,
}
`;

exports[`should parse digitableLine properly 9`] = `
Object {
  "amount": 743.84,
  "barCode": "85800000007438403281926307201925230472960052",
  "expirationDate": null,
}
`;
